errors.go

```diff
@@ -0,0 +1,207 @@
+// are supplied to a command.
+type InvalidArgCountError struct {
+	cmd     *Command
+	args    []string
+	atLeast int
+	atMost  int
+}
+
+// Error implements error.
+func (e *InvalidArgCountError) Error() string {
+	if e.atMost == -1 && e.atLeast >= 0 { // MinimumNArgs
+		return fmt.Sprintf("requires at least %d arg(s), only received %d", e.atLeast, len(e.args))
+	}
+
+	if e.atLeast == -1 && e.atMost >= 0 { // MaximumNArgs
+		return fmt.Sprintf("accepts at most %d arg(s), received %d", e.atMost, len(e.args))
+	}
+
+	if e.atLeast == e.atMost && e.atLeast != -1 { // ExactArgs
+		return fmt.Sprintf("accepts %d arg(s), received %d", e.atLeast, len(e.args))
+	}
+
+	// RangeArgs
+	return fmt.Sprintf("accepts between %d and %d arg(s), received %d", e.atLeast, e.atMost, len(e.args))
+}
+
+// GetCommand returns the Command that the error occurred in.
+func (e *InvalidArgCountError) GetCommand() *Command {
+	return e.cmd
+}
+
+// GetArguments returns the invalid arguments.
+func (e *InvalidArgCountError) GetArguments() []string {
+	return e.args
+}
+
+// GetMinArgumentCount returns the minimum (inclusive) number of arguments
+// that the command requires. If there is no minimum, a value of -1 is returned.
+func (e *InvalidArgCountError) GetMinArgumentCount() int {
+	return e.atLeast
+}
+
+// GetMaxArgumentCount returns the maximum (inclusive) number of arguments
+// that the command requires. If there is no maximum, a value of -1 is returned.
+func (e *InvalidArgCountError) GetMaxArgumentCount() int {
+	return e.atMost
+}
+
+// InvalidArgCountError is the error returned an invalid argument is present.
+type InvalidArgValueError struct {
+	cmd         *Command
+	arg         string
+	suggestions string
+}
+
+// Error implements error.
+func (e *InvalidArgValueError) Error() string {
+	return fmt.Sprintf("invalid argument %q for %q%s", e.arg, e.cmd.CommandPath(), e.suggestions)
+}
+
+// GetCommand returns the Command that the error occurred in.
+func (e *InvalidArgValueError) GetCommand() *Command {
+	return e.cmd
+}
+
+// GetArgument returns the invalid argument.
+func (e *InvalidArgValueError) GetArgument() string {
+	return e.arg
+}
+
+// GetSuggestions returns suggestions, if there are any.
+func (e *InvalidArgValueError) GetSuggestions() string {
+	return e.suggestions
+}
+
+// UnknownSubcommandError is the error returned when a subcommand can not be
+// found.
+type UnknownSubcommandError struct {
+	cmd         *Command
+	subcmd      string
+	suggestions string
+}
+
+// GetCommand returns the Command that the error occurred in.
+func (e *UnknownSubcommandError) GetCommand() *Command {
+	return e.cmd
+}
+
+// GetSubcommand returns the unknown subcommand name.
+func (e *UnknownSubcommandError) GetSubcommand() string {
+	return e.subcmd
+}
+
+// GetSuggestions returns suggestions, if there are any.
+func (e *UnknownSubcommandError) GetSuggestions() string {
+	return e.suggestions
+}
```

Comment:
Here an for other errors, I would have preferred if this method returned a []string

But it would require to rewrite findSuggestions, or at least to split by \n here, but it would be ugly and dirty

Reply:
That's totally fair. I added [another commit](https://github.com/spf13/cobra/pull/2266/commits/8dd7f633ac91a8fe286b9d6d1ff5d2471eb624b6) to split up `findSuggestions` into two separate functions, and I changed the error structs to hold `suggestions []string` and return the suggestions slice instead of the merged string.